src/parser.rs

```diff
@@ -3,9 +3,59 @@ use error::Error;
 
 use nom::types::CompleteStr;
 use nom::{eol, space, space0};
+use nom_locate::LocatedSpan;
 
+use std::fmt;
 use std::num;
 
+pub type Span<'a> = LocatedSpan<CompleteStr<'a>>;
+
+pub enum ErrorCode {
+    ExpectedStringExpression,
+    Unknown,
+}
+
+impl From<u32> for ErrorCode {
+    fn from(code: u32) -> Self {
+        match code {
+            0 => ErrorCode::ExpectedStringExpression,
+            _ => ErrorCode::Unknown,
+        }
+    }
+}
+
+impl fmt::Display for ErrorCode {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ErrorCode::ExpectedStringExpression => f.write_str("string expression expected"),
+            ErrorCode::Unknown => f.write_str("unknown"),
+        }
+    }
+}
+
+impl ErrorCode {
+    /// Converts error code to BASIC conform error output.
+    ///
+    /// # Arguments
+    ///
+    /// `line` - Full line string where the error happened.
+    /// `remaining` - Remaining statement which could not be parsed. Must be a substring of `line`.
+    pub fn to_string(&self, line: &str, remaining: &str) -> Option<String> {
+        let mut parts = line.trim().splitn(2, ' ');
+        let line_number = parts.next()?;
```

Comment:
I don't get this. I thought we do know the line number now. Why do we have to get this from the line string?

Reply:
`position.line` is the line number of the current source code line, not the line number label of BASIC. I think in the future we should rename BASIC's line number to line label.

Reply:
Good point.

Reply:
`line_label` would slightly deviate from the grammar definition in ECMA, however this should not be a problem in `ast`. There it us up to us how to call stuff.